src/utilities/flattenRoutes.ts

```diff
@@ -1,31 +1,39 @@
-import { RouteFlat, Routes, isParentRoute, isNamedRoute } from '@/types'
+import { RouteFlat, Routes, isParentRoute, isNamedRoute, Path } from '@/types'
+import { mergeParams, path } from '@/utilities'
 
-export function flattenRoutes(routes: Routes, path = ''): RouteFlat[] {
+export function flattenRoutes(routes: Routes, parentPath = '', parentParams = {}): RouteFlat[] {
   return routes.reduce<RouteFlat[]>((value, route) => {
-    const fullPath = path + route.path
+    const { params } = toPath(route.path)
+    const fullPath = parentPath + route.path.toString()
+    const fullParams = mergeParams(parentParams, params)
 
     if (isParentRoute(route)) {
-      const flattened = flattenRoutes(route.children, fullPath)
+      const flattened = flattenRoutes(route.children, fullPath, fullParams)
 
       value.push(...flattened)
     }
 
     if (isNamedRoute(route)) {
       value.push({
         name: route.name,
-        regex: generateRouteRegexPattern(fullPath),
         path: fullPath,
+        params: fullParams,
+        regex: generateRouteRegexPattern(fullPath),
       })
     }
 
     return value
   }, [])
 }
 
-export function generateRouteRegexPattern(path: string): RegExp {
+export function toPath(value: string | Path): Path {
+  return typeof value === 'string' ? path(value, {}) : value
+}
+
+export function generateRouteRegexPattern(value: string): RegExp {
   const optionalParamRegex = /(:\?[\w]+)(?=\W|$)/g
   const requiredParamRegex = /(:[\w]+)(?=\W|$)/g
 
-  const routeRegex = path.replace(optionalParamRegex, '([^/]*)').replace(requiredParamRegex, '([^/]+)')
+  const routeRegex = value.replace(optionalParamRegex, '(.*)').replace(requiredParamRegex, '(.+)')
```

Comment:
In the path utility there's a single regex that matches both optional and required params. Can we reuse that here? Quite possible they are incompatible.

Reply:
that's a good catch, but there is a subtle difference in the replacement values.

Notice how the optional gets replaced with `(.*)` (any char 0 or more times)
and the required gets replaced with `(.+)` (any char 1 or more times)

Reply:
ah right 👍